activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -64,6 +66,34 @@ def find_finder_class_for(record)
         class_hierarchy.detect { |klass| !klass.abstract_class? }
       end
 
+      def validation_needed?(klass, record, attribute)
+        return true if options[:conditions] || options.key?(:case_sensitive)
+
+        scope = Array(options[:scope]).map(&:to_s)
+        attributes = scope + [attribute.to_s]
+
+        return true if attributes.any? { |attribute| record.class._reflect_on_association(attribute) }
+        return true if attributes.any? { |attribute| record.attribute_changed?(attribute) ||
+                                                     record.read_attribute(attribute).nil? }
+
+        !covered_by_unique_index?(klass, attribute, scope)
+      end
+
+      def covered_by_unique_index?(klass, attribute, scope)
+        return @covered_by_unique_index[attribute] if defined?(@covered_by_unique_index)
+
+        @covered_by_unique_index = self.attributes.index_with do |attribute|
+          attributes = scope + [attribute.to_s]
+          klass.connection.schema_cache.indexes(klass.table_name).any? do |index|
+            index.unique &&
+              index.where.nil? &&
+              (index.columns - attributes).empty?
+          end
+        end
+
+        @covered_by_unique_index[attribute]
```

Comment:
```suggestion
        @covered_by_unique_index[attribute.to_s]
```

If you call `to_s` in the generation code, you need to call it here as well.